consensus/event_loop.go

```diff
@@ -0,0 +1,908 @@
+	fc.status = eventLoopStatusInitialized
+
+	return nil
+}
+
+// GetEvents returns the event loop's external channel for signaling. We need a getter function
+// to ensure that this struct implements the FastHotStuffEventLoop interface type.
+func (fc *fastHotStuffEventLoop) GetEvents() chan *FastHotStuffEvent {
+	return fc.Events
+}
+
+// GetCurrentView is a simple getter that returns the event loop's current view. It does not need
+// to be thread-safe. The caller is expected to use it in a thread-safe manner, at a time when
+// the view is guaranteed to not change.
+func (fc *fastHotStuffEventLoop) GetCurrentView() uint64 {
+	return fc.currentView
+}
+
+// AdvanceViewOnTimeout is called when the tip has not changed but the event loop has timed out. This
+// function advances the view and resets the crank timer and timeout scheduled tasks.
+func (fc *fastHotStuffEventLoop) AdvanceViewOnTimeout() (uint64, error) {
+	// Grab the event loop's lock
+	fc.lock.Lock()
+	defer fc.lock.Unlock()
+
+	// Ensure the event loop is running. This guarantees that the chain tip and validator list
+	// have already been set.
+	if fc.status != eventLoopStatusRunning {
+		return 0, errors.New("FastHotStuffEventLoop.AdvanceViewOnTimeout: Event loop is not running")
+	}
+
+	// Advance the view
+	fc.currentView++
+
+	// Reset QC construction status for the current view
+	fc.hasCrankTimerRunForCurrentView = false
+	fc.hasConstructedQCInCurrentView = false
+
+	// Evict all stale votes and timeouts
+	fc.evictStaleVotesAndTimeouts()
+
+	// Schedule the next crank timer and timeout scheduled tasks
+	fc.resetScheduledTasks()
+
+	return fc.currentView, nil
+}
+
+// ProcessTipBlock must only be called when the server has accepted a new block, connected it
+// to the tip of the blockchain, and determined that the block is safe to vote on. Given such a
+// block, this function resets internal state and schedules the next crank timer and timeout
+// timer.
+//
+// Expected params:
+//   - tip: the current uncommitted tip of the blockchain, with the validator list at that block height
+//   - safeBlocks: an unordered slice of blocks including the committed tip, the uncommitted tip,
+//     all ancestors of the uncommitted tip that are safe to extend from, and all blocks from forks
+//     that are safe to extend from. This function does not validate the collection of blocks. It
+//     expects the server to know and decide what blocks are safe to extend from.
+func (fc *fastHotStuffEventLoop) ProcessTipBlock(tip BlockWithValidatorList, safeBlocks []BlockWithValidatorList) error {
+	// Grab the event loop's lock
+	fc.lock.Lock()
+	defer fc.lock.Unlock()
+
+	// Ensure the event loop is running
+	if fc.status != eventLoopStatusRunning {
+		return errors.New("FastHotStuffEventLoop.ProcessTipBlock: Event loop is not running")
+	}
+
+	// Validate the safe blocks and validator lists, and store them
+	if err := fc.storeBlocks(tip, safeBlocks); err != nil {
```

Comment:
Consider adding an error check before this point to verify that fc.tip.GetView() <= tip.GetView(). As it stands right now, I think you could call ProcessTipBlock with a stale tip and it would accept it, which I don't think is desired behavior.

Reply:
Then again maybe you want this function to "reset" everything, irrespective of what the prior state was. If that's the case then it's fine as is.

Reply:
Correct. We want this function to reset everything regardless of what the previous state was. The consensus rules to decide on the next tip are all implemented in ProcessBlockPoS. We avoid duplicating them here and instead trust the new tip passed by the caller. If ProcessBlockPoS decides a new block is the new tip, we accept it at face value here. If for whatever reason we decide to update the view in ProcessBlockPoS, this consensus module just defers to that reasoning. It's mostly for separation of concerns.

In practice though, the view will never regress.